packages/devtools_shared/lib/src/server/server_api.dart

```diff
@@ -260,9 +263,12 @@ class ServerApi {
           deeplinkManager,
         );
       case DtdApi.apiGetDtdUri:
-        return _encodeResponse(
-          {DtdApi.uriPropertyName: dtdUri},
-          api: api,
+        return _DtdApiHandler.handleGetDtdUri(api, dtd);
+      case DtdApi.apiSetDtdWorkspaceRoots:
+        return _DtdApiHandler.handleSetDtdWorkspaceRoots(
```

Comment:
Is there a way we can protect this with some sort of key that only the DevTools web app has? I guess implementing a "trusted client" for DevTools server?

Reply:
Filed https://github.com/dart-lang/sdk/issues/55074 to track that work. As discussed offline, I have restructured this PR so that DevTools server is the one looking up the root for a VM service connection and setting the roots on DTD.